firedrake/adjoint/variational_solver.py

```diff
@@ -64,3 +76,41 @@ def wrapper(self, **kwargs):
             return out
 
         return wrapper
+
+    @no_annotations
+    def _ad_problem_clone(self, problem, dependencies):
+        """Replaces every coefficient in the residual and jacobian with a deepcopy to return
+        a clone of the original NonlinearVariationalProblem instance. We'll be modifying the
+        numerical values of the coefficients in the residual and jacobian, so in order not to
+        affect the user-defined self._ad_problem.F, self._ad_problem.J and self._ad_problem.u
+        expressions, we'll instead create clones of them.
+        """
+        from firedrake import NonlinearVariationalProblem
+        F_replace_map = {}
+        J_replace_map = {}
+
+        F_coefficients = problem.F.coefficients()
+        J_coefficients = problem.J.coefficients()
+
+        for block_variable in dependencies:
+            coeff = block_variable.output
+            if coeff in F_coefficients and coeff not in F_replace_map:
+                if isinstance(coeff, Constant):
+                    F_replace_map[coeff] = copy.deepcopy(coeff)
+                else:
+                    F_replace_map[coeff] = coeff.copy(deepcopy=True)
+                F_replace_map[coeff]._ad_count = coeff.count()
```

Comment:
Here, you are creating a new attribute for `Coefficients`, I think it would be better to define this new attribute directly in the appropriate class. I guess if it is "adjoint-related" you could use the `FunctionMixin` (cf. `/firedrake/adjoint/function.py`).

Also if I have well understood, here you replace the coefficients turning up in the forms (F and J) with copies, and in order to keep track of this substitution, you attach the count of the original coefficients to their copies introducing a new attribute (`_ad_count`) in order to identify the copies later on (in the block) to then replace them with the saved outputs, is it correct?

How about instead attaching this information (mapping) to the `NonlinearVariationalProblem` itself (using for example the `NonlinearVariationalProblemMixin`) and transmit it to the block, so that you have it when needed (i.e. in `_forward_solve`)?

Reply:
Your understanding of the use the attribute `_ad_count` is absolutely correct. I think creating a `NonlinearVariationalProblemMixin` class and attaching the mapping to it seems reasonable.

Reply:
Ok perfect, actually the `NonlinearVariationalProblemMixin` class already exists (cf. `firedrake/adjoint/variational_solver.py`)